network.go

```diff
@@ -50,137 +40,181 @@ func (nw *netWatcher) Results() (MeasurementsMap, error) {
+			return true
+		}
+	}
 
-			if !isExcluded {
-				for _, excludedIf := range nw.cagent.NetInterfaceExclude {
-					if strings.EqualFold(netIf.Name, excludedIf) {
-						isExcluded = true
-						break
-					}
-				}
-			}
+	return false
+}
 
-			if !isExcluded {
-				for _, re := range netInterfaceExcludeRegexCompiled {
-					if re.MatchString(netIf.Name) {
-						isExcluded = true
-						break
-					}
-				}
-			}
+func isInterfaceDown(netIf *utilnet.InterfaceStat) bool {
+	for _, flag := range netIf.Flags {
+		if flag == "up" {
+			return false
+		}
+	}
+
+	return true
+}
 
-			nw.ExcludedInterfaceCache[netIf.Name] = isExcluded
+func (nw *netWatcher) isInterfaceExcludedByName(netIf *utilnet.InterfaceStat) bool {
+	for _, excludedIf := range nw.cagent.NetInterfaceExclude {
+		if strings.EqualFold(netIf.Name, excludedIf) {
+			return true
+		}
+	}
+	return false
+}
 
-			if isExcluded {
+func (nw *netWatcher) isInterfaceExcludedByRegexp(netIf *utilnet.InterfaceStat) bool {
+	for _, re := range nw.InterfaceExcludeRegexCompiled() {
+		if re.MatchString(netIf.Name) {
+			return true
+		}
+	}
+	return false
+}
+
+func (nw *netWatcher) ExcludedInterfacesByNameMap(allInterfaces []utilnet.InterfaceStat) map[string]struct{} {
+	excludedInterfaces := map[string]struct{}{}
+
+	for _, netIf := range allInterfaces {
+		// use a cache for excluded interfaces, because all the checks(except UP/DOWN state) are constant for the same interface&config
+		if isExcluded, cacheExists := nw.constantlyExcludedInterfaceCache[netIf.Name]; cacheExists {
+			if isExcluded ||
+				nw.cagent.NetInterfaceExcludeDisconnected && isInterfaceDown(&netIf) {
+				// interface is found excluded in the cache or has a DOWN state
+				excludedInterfaces[netIf.Name] = struct{}{}
 				log.Debugf("[NET] interface excluded: %s", netIf.Name)
 				continue
 			}
+		} else {
+			if nw.cagent.NetInterfaceExcludeLoopback && isInterfaceLoobpack(&netIf) ||
+				nw.isInterfaceExcludedByName(&netIf) ||
+				nw.isInterfaceExcludedByRegexp(&netIf) {
+				// add the excluded interface to the cache because this checks are constant
+				nw.constantlyExcludedInterfaceCache[netIf.Name] = true
+			} else if nw.cagent.NetInterfaceExcludeDisconnected && isInterfaceDown(&netIf) {
+				// exclude DOWN interface for now
+				// lets cache it as false and then we will only check UP/DOWN status
+				nw.constantlyExcludedInterfaceCache[netIf.Name] = false
+			} else {
+				// interface is not excluded
+				nw.constantlyExcludedInterfaceCache[netIf.Name] = false
+				continue
+			}
+
+			excludedInterfaces[netIf.Name] = struct{}{}
+			log.Debugf("[NET] interface excluded: %s", netIf.Name)
 		}
 	}
+	return excludedInterfaces
+}
 
-	ctx, _ = context.WithTimeout(context.Background(), fsGetUsageTimeout)
-	counters, err := utilnet.IOCountersWithContext(ctx, true)
+func (nw *netWatcher) fillEmptyMeasurements(results MeasurementsMap, interfaces []utilnet.InterfaceStat, excludedInterfacesByName map[string]struct{}) {
```

Comment:
please add some comment that explain what and why is filled here. I only see `= nil` and this doesn't fill something in my understanding 🤔

Reply:
I made some comment, is it clear enough?

Reply:
Ah I see, it creates all the keys inside the map